File: Bronze/2017/February/02_why_did_the_cow_cross_the_road_ii/stack/WhyDidTheCowCrossTheRoadII2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class WhyDidTheCowCrossTheRoadII2 {
    static Stack<Character> stack = new Stack<>();
    static int[] visited = new int[26];
    static int ans = 0;
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("circlecross.in"));
        PrintWriter pw = new PrintWriter("circlecross.out");

        String s = r.readLine();

        for (int i = 0; i < 52; i++) {
            char ch = s.charAt(i);
            int chIndex = ch - 'A';
            if (visited[chIndex] == 0) {
                visited[chIndex] = 1;
                stack.push(ch);
            } else {
                int startPos = 0;
                while (startPos < stack.size()) {
                    if (stack.get(startPos) == ch) {
                        break;
                    }
                    startPos++;
                }
                ans += stack.size() - startPos - 1;
                stack.remove(startPos);
            }
        }

        pw.println(ans);
        pw.close();
    }
}

File: Bronze/2017/February/02_why_did_the_cow_cross_the_road_ii/stack/why_did_the_cow_cross_the_road_ii2.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int main() {
    ifstream fin("circlecross.in");
    ofstream fout("circlecross.out");

    string s;
    fin >> s;

    vector<char> char_stack;
    int visited[26] = {0};
    int ans = 0;

    for (int i = 0; i < 52; i++) {
        char ch = s[i];
        int ch_index = ch - 'A';
        if (visited[ch_index] == 0) {
            visited[ch_index] = 1;
            char_stack.push_back(ch);
        } else {
            int start_pos = 0;
            while (start_pos < char_stack.size()) {
                if (char_stack[start_pos] == ch) {
                    break;
                }
                start_pos++;
            }
            ans += char_stack.size() - start_pos - 1;
            char_stack.erase(char_stack.begin() + start_pos);
        }
    }

    fout << ans << endl;
    fin.close();
    fout.close();

    return 0;
}
